File: android/app/src/main/kotlin/com/usernode_labs/usernode/session/SessionAuthorityNative.kt
package com.usernode_labs.usernode.session

import android.content.Context
import com.usernode_labs.usernode.alarm.RuntimeOwner
import java.io.File

/** Thin, read-only Android client for the Rust-owned process authority. */
internal object SessionAuthorityNative {
    private const val DIRECTORY_NAME = "session-authority"

    private val libraryLoaded = try {
        System.loadLibrary("usernode")
        true
    } catch (_: UnsatisfiedLinkError) {
        false
    }

    private external fun readAdmissionJson(journalDirectory: String): String

    private external fun admitsBackgroundRuntime(
        journalDirectory: String,
        sessionId: String,
        runtimeGeneration: Long,
        accountId: String,
        address: String,
    ): Boolean

    fun admissionJson(context: Context): String {
        if (!libraryLoaded) {
            return """{"status":"terminal","reason":"Rust authority library unavailable"}"""
        }
        return readAdmissionJson(journalDirectory(context))
    }

    fun isBackgroundRuntimeAdmitted(
        context: Context,
        owner: RuntimeOwner,
    ): Boolean {
        if (!libraryLoaded) {
            return false
        }
        return admitsBackgroundRuntime(
            journalDirectory(context),
            owner.sessionId,
            owner.runtimeGeneration,
            owner.accountId,
            owner.address,
        )
    }

    private fun journalDirectory(context: Context): String =
        File(context.filesDir, DIRECTORY_NAME).absolutePath
}

File: ios/Runner/SessionAuthorityNative.swift
import Foundation

/// Thin, read-only iOS client for the Rust-owned process authority.
enum SessionAuthorityNative {
  private static let directoryName = "session-authority"

  static func admissionJSON() -> String {
    guard let directory = journalDirectory() else {
      return #"{"status":"terminal","reason":"Application Support unavailable"}"#
    }
    let value = directory.path.withCString {
      usernode_session_authority_admission_json($0)
    }
    guard let value else {
      return #"{"status":"terminal","reason":"Rust authority returned no value"}"#
    }
    defer { usernode_session_authority_string_free(value) }
    return String(cString: value)
  }

  private static func journalDirectory() -> URL? {
    try? FileManager.default
      .url(
        for: .applicationSupportDirectory,
        in: .userDomainMask,
        appropriateFor: nil,
        create: false
      )
      .appendingPathComponent(directoryName, isDirectory: true)
  }
}
